### PR TITLE
refactor: fill in basic types in rules.ts

### PR DESCRIPTION
**Pull request by UnderKoen**

### CODE REVIEW

File: packages/loot-core/src/server/accounts/rules.ts
Comment (jfdoming) [inline packages/loot-core/src/server/accounts/rules.ts:733]: From looking through the code-base, I _think_ this would be the type?
```suggestion
  stage: 'pre' | 'default' | 'post';
```

File: packages/loot-core/src/server/accounts/rules.ts
Comment (jfdoming) [inline packages/loot-core/src/server/accounts/rules.ts:26]: Wow I didn't know about `asserts`, this is great!

Comment (jfdoming) [reply packages/loot-core/src/server/accounts/rules.ts:26]: Wondering your thoughts on doing `test: boolean` and explicitly converting at call sites? My thinking:
- Pro: this function is better documented, as `test` is always coerced into a boolean
- Con: have to modify all call sites and do minor unnecessary conversion

File: packages/loot-core/src/server/accounts/rules.ts
Comment (jfdoming) [inline packages/loot-core/src/server/accounts/rules.ts:835]: Looking at `rule.d.ts`, it seems like the value is always a `string`, `number`, `string[]`, or `number[]`. What do you think about either explicitly enumerating those types here, or using `BaseConditionEntity<keyof FieldValueTypes, RuleConditionOp>['value']`? The latter one would be nice since it would support any new types we add in the future.

File: packages/loot-core/src/server/accounts/rules.ts
Comment (jfdoming) [inline packages/loot-core/src/server/accounts/rules.ts:882]:
```suggestion
  getApplicableRules(object: Record<string, unknown>): Set<Rule> {
```
(or `Record<string, string | number | string[] | number[]>` as the case may be)

**Review by jfdoming — COMMENTED** [review]
Thanks for working on this!

File: packages/loot-core/src/server/accounts/rules.ts
Comment (UnderKoen) [reply packages/loot-core/src/server/accounts/rules.ts:733]: I though this was `'pre' | 'post' | null`

File: packages/loot-core/src/server/accounts/rules.ts
Comment (UnderKoen) [reply packages/loot-core/src/server/accounts/rules.ts:26]: I would think it's unnecessary. `assert(!!value, ...)` is not much clearer than `assert(value, ...`.

File: packages/loot-core/src/server/accounts/rules.ts
Comment (UnderKoen) [reply packages/loot-core/src/server/accounts/rules.ts:835]: The function doesn't realy care what it is passed. So I think leaving it broarder work better?

File: packages/loot-core/src/server/accounts/rules.ts
Comment (UnderKoen) [reply packages/loot-core/src/server/accounts/rules.ts:882]: I was working on making the whole class generic. So that the type is really enforced. 

The problem I had was that it makes thinks some more verbose while Rule are currently only be used for transactions.  And a lot of the logic inside the file depends on it being a TransactionEnity. 

So I didn't know if it was worth adding.

If you would like to look: https://github.com/UnderKoen/actual/commit/4af6fc8402f83b569fec8dc9a893236cf8e20099

Comment (jfdoming) [reply packages/loot-core/src/server/accounts/rules.ts:882]: Ah, I see. Let's defer to a future PR then!

File: packages/loot-core/src/server/accounts/rules.ts
Comment (jfdoming) [reply packages/loot-core/src/server/accounts/rules.ts:26]: fair enough!

File: packages/loot-core/src/server/accounts/rules.ts
Comment (jfdoming) [reply packages/loot-core/src/server/accounts/rules.ts:733]: Oh you're right based on my reading of the logic. I think the place I looked is actually a bug. Mind updating the logic [here](https://github.com/actualbudget/actual/blob/master/packages/desktop-client/src/components/accounts/Account.tsx#L1193) to reflect? And maybe good to fix the type [here](https://github.com/actualbudget/actual/blob/master/packages/loot-core/src/types/models/rule.d.ts#L4)

File: packages/loot-core/src/server/accounts/rules.ts
Comment (jfdoming) [reply packages/loot-core/src/server/accounts/rules.ts:835]: Sure, it probably doesn't matter much in this case

File: packages/loot-core/src/types/models/rule.d.ts
Comment (UnderKoen) [reply packages/loot-core/src/types/models/rule.d.ts:4]: `cleanup` and `modify` are legacy and are changed in that function to `pre` so the type is incorrect but ensures good usage

**Review by MikesGlitch — COMMENTED** [review]
This is a nice improvement! Sorry it took so long to get a review. 

Can you resolve the conflict? I'm happy to approve this once done 👍

**Review by MikesGlitch — APPROVED** [review]
Nice work!

**Pull request merged**
